explainer.md

```diff
@@ -68,356 +107,405 @@ If the set of pages being badged corresponds to an [Installed Web Application](h
+notification, or turn off low-priority messages.
+
+#### A separate channel for Badge payloads
+
+A more comprehensive solution includes some significant additions to the Push
+API (which the authors of the Badge spec are not really equipped to do).
+
+(This has been briefly discussed with Peter Beverloo, an editor of the Push API
+spec, but only at a very high level.)
+
+Instead of allowing the service worker to run JavaScript code and call the
+`Badge` API, we would introduce a new concept to a Push subscription called a
+"channel", which dictates what is done with the payload upon receipt. The
+default channel would be "event" (the payload is delivered to the `"push"`
+event), with a new channel, "badge", which imposes a specific format to the
+payload. The payload would now be interpreted as a JSON dictionary containing
+parameters to the `Badge.set` API. Upon receipt of the push, the user agent
+would automatically call the `Badge` API without running any user code, and with
+no requirement to show a notification.
+
+This solution still has the following two flaws (expressed by Peter Beverloo):
+
+1. The server can still use badge delivery to know whether device is online
+   through [delivery receipts](https://tools.ietf.org/html/rfc8030#section-6.3).
+   This may or may not be a privacy concern.
+2. (On mobile) (at least on Android and iOS) the browser needs to be woken up to
+   decrypt and process the payload; it can't be processed by the host OS's push
+   receipt system due to the encryption. This could be a big resource drain if
+   badges are pushed too frequently (the same concern applies for notifications,
+   but as above, there is a natural tendency for developers to reduce the number
+   of user-visible notifications due to user spam; the same forcing function
+   does not apply for setting a badge).
+
+#### Conclusion
+
+Both of the above changes present huge obstacles and discussions involving
+privacy, resource usage, and utility trade-offs. Due to the increased
+complexity, we are not considering changes to the Push API at this time.
+
+## Feature detection
+
+Ideally we would not provide any feedback to the site as to how the badge is
+going to be displayed to the user, since it's a user interface concern that
+should be at the discretion of the user agent. For example, the site shouldn't
+need to know whether the badge is displayed on a tab icon or on an app icon.
+
+However, a practical concern overrides this: because the favicon can be (and
+often is) used to display a page badge, sites upgrading to the `Badge` API need
+to know whether they need to fall back to setting the favicon (since,
+presumably, they don't want the badge being displayed twice). So we need to
+provide a way to feature detect — not just "whether the Badge API is supported",
+but "whether a Badge-API badge will show up on or near the favicon". So we
+introduce the `Badge.canBadgeDocument` API:
```

Comment:
What about having `Badge.set()` return a list of places that were badged? 

or `Badge.setDocument(4)` returns a boolean success status?

Reply:
That's probably better.

Reply:
Discussed doing this separately offline.